cpu/cc2538/periph/i2c.c

```diff
@@ -8,658 +9,331 @@
 
-    return n;
-}
-
-int i2c_write_reg(i2c_t dev, uint8_t address, uint8_t reg, uint8_t data)
-{
-    return i2c_write_regs(dev, address, reg, &data, 1);
+    return 0;
 }
 
-int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, const void *data, int length)
+int i2c_write_bytes(i2c_t dev, uint16_t addr, const void *data,
+                    size_t len, uint8_t flags)
 {
-    uint_fast8_t stat;
-    const uint8_t *my_data = data;
+    DEBUG("%s\n", __FUNCTION__);
+    DEBUG(" - I2C master status (%u).\n", _i2c_master_status());
+    (void)flags;
 
-    if (dev != I2C_0) {
-        return -1;
+    if ((dev >= I2C_NUMOF) || (data == NULL) || (len == 0)) {
+        return -EINVAL;
     }
-
-    /* Transmit reg byte to slave */
-    if (i2c_busy()) {
-        return 0;
+    if (_i2c_master_busy()) {
+        DEBUG("i2c_write_bytes: device busy!\n");
+        return -EAGAIN;
     }
 
-    WARN_IF(I2CM_STAT & BUSBSY);
+    /* set slave address for write */
+    _i2c_master_slave_addr(addr, false);
 
-    if (I2CM_STAT & BUSBSY) {
-        recover_i2c_bus();
+    const uint8_t *buf = data;
 
-        if (I2CM_STAT & BUSBSY) {
-            return 0;
+    for (size_t n = 0; n < len; n++) {
+        uint_fast8_t cmd = RUN;
+        if ((n == 0) && !(flags & I2C_NOSTART)) {
+            cmd |= START;
         }
-    }
-
-    I2CM_SA = address << 1;
-    I2CM_DR = reg;
-
-    uint_fast8_t flags = (length > 0)? (START | RUN) : (STOP | START | RUN);
-    stat = i2c_ctrl_blocking(flags);
-
-    if (stat & ARBLST) {
-        return 0;
-    }
-    else if (stat & ANY_ERROR) {
-        i2cm_ctrl_write(STOP);
-        return 0;
-    }
-    else {
-        /* Transmit data to slave */
-        int n = 0;
-
-        flags &= ~START;
-
-        for (n = 0; n < length; n++) {
-            if (n >= length - 1) flags |= STOP;
-            WARN_IF(I2CM_STAT & BUSY);
-            I2CM_DR = my_data[n];
-
-            i2c_ctrl_blocking(flags);
-
-            WARN_IF(I2CM_STAT & ARBLST);
-            WARN_IF(I2CM_STAT & DATACK);
-            WARN_IF(I2CM_STAT & ADRACK);
-            WARN_IF(I2CM_STAT & ERROR);
-
-            if (I2CM_STAT & ARBLST) {
-                 break;
+        if (((len - n) == 1) && !(flags & I2C_NOSTOP)) {
+            cmd |= STOP;
+        }
+        /* write byte to mem */
+        _i2c_master_data_put(buf[n]);
+        /* run command */
+        _i2c_master_ctrl(cmd);
+        /* wait until master is done transferring */
+        DEBUG ("%s: wait for master...\n", __FUNCTION__);
+        unsigned cw = CMD_WAIT;
+        while (_i2c_master_busy() || (cw--)) {}
+
+        /* check master status */
+        uint_fast8_t stat = _i2c_master_status();
+        DEBUG ("%s: I2C master status (%u).\n", __FUNCTION__, stat);
+        if (stat & ANY_ERROR) {
+            _i2c_master_ctrl(STOP);
+            DEBUG("\tI2C master error: ");
+            if (stat &  DATACK) {
+                DEBUG("data ack lost!\n");
+                return -EIO;
```

Comment:
If you can find a way to clear the error flags before exiting that would be nice... I ran out of time.

Reply:
The only way I could clear the flag is with a successful write.  Maybe try the standard solution?  (turn it off and on again)